nabs/streams.py

```diff
@@ -0,0 +1,73 @@
+"""Standard data streams for automated routines"""
+import numpy as np
+from bluesky.callbacks.stream import LiveDispatcher
+
+
+class AverageStream(LiveDispatcher):
+    """
+    Stream that averages data points together
+
+    As event documents are emitted from the RunEngine they are collected by
+    AverageStream, averaged and re-emitted as a secondary event stream. This
+    allows callbacks to subscribe to the average stream for more elegant
+    visualization and analysis.
+
+    The average stream can either be constructed with a specific number of
+    points to average or this information can be passed through in the start
+    document metadata. See the :meth:`.start` for more information.
+
+    Parameters
+    ----------
+    n : int, optional
+        Number of points to average together
+    """
+    def __init__(self, n=None):
+        self.n = n
+        self.last_event = None
+        self.raw_cache = list()
+        super().__init__()
+
+    def start(self, doc):
+        """
+        Create the stream after seeing the start document
+
+        The callback looks for the 'average' key in the start document to
+        configure itself.
+        """
+        # Grab the average key
+        self.n = doc.get('average', self.n)
```

Comment:
This is an interesting idea, I haven't seen any code take action based on start doc metadata before. This doesn't appear usable in `measure_average` because num is a required argument, but I'm assuming the point is to leave this here for use later?

Reply:
This was a suggestion from the NSLS team when we were there. I liked the idea and left it in though I think taking it to the extreme could be scary. The crucial thing that could break this plan is that the callback is out of sync configuration-wise, giving multiple ways to sync things seems reasonable.